Remotion/Data/DomainObjects/Persistence/Rdbms/SqlServer/DbCommandBuilders/Specifications/SqlTableValuedParameterComparedColumnSpecification.cs

```diff
@@ -0,0 +1,111 @@
+// Copyright (c) rubicon IT GmbH, www.rubicon.eu
+// 
+// The re-motion Core Framework is free software; you can redistribute it 
+// and/or modify it under the terms of the GNU Lesser General Public License 
+// as published by the Free Software Foundation; either version 2.1 of the 
+// License, or (at your option) any later version.
+// 
+// re-motion is distributed in the hope that it will be useful, 
+// but WITHOUT ANY WARRANTY; without even the implied warranty of 
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the 
+// GNU Lesser General Public License for more details.
+// 
+// You should have received a copy of the GNU Lesser General Public License
+// along with re-motion; if not, see http://www.gnu.org/licenses.
+//
+using System;
+using System.Collections.Generic;
+using System.Data;
+using System.Data.SqlClient;
+using System.Linq;
+using System.Text;
+using Microsoft.SqlServer.Server;
+using Remotion.Data.DomainObjects.Persistence.Rdbms.DbCommandBuilders.Specifications;
+using Remotion.Data.DomainObjects.Persistence.Rdbms.Model;
+using Remotion.Data.DomainObjects.Persistence.Rdbms.SqlServer.Parameters;
+using Remotion.Utilities;
+
+namespace Remotion.Data.DomainObjects.Persistence.Rdbms.SqlServer.DbCommandBuilders.Specifications;
+
+/// <summary>
+/// Specifies a set of <see cref="ObjectValues"/> that are transmitted as a table-valued parameter, to be used in a WHERE-IN-clause.
+/// </summary>
+public class SqlTableValuedParameterComparedColumnSpecification : IComparedColumnsSpecification
+{
+  /// <summary>
+  /// The column whose values are looked up in the <see cref="ObjectValues"/>.
+  /// </summary>
+  public ColumnDefinition ColumnDefinition { get; }
+
+  /// <summary>
+  /// The values that fulfill the filter criterion.  
+  /// </summary>
+  public IEnumerable<object?> ObjectValues { get; }
+
+  public SqlTableValuedParameterComparedColumnSpecification (ColumnDefinition columnDefinition, IEnumerable<object?> objectValues)
+  {
+    ArgumentUtility.CheckNotNull(nameof(columnDefinition), columnDefinition);
+    ArgumentUtility.CheckNotNull(nameof(objectValues), objectValues);
+
+    ColumnDefinition = columnDefinition;
+    ObjectValues = objectValues;
+  }
+
+  /// <inheritdoc />
+  public void AddParameters (IDbCommand command, ISqlDialect sqlDialect)
+  {
+    ArgumentUtility.CheckNotNull(nameof(command), command);
+    ArgumentUtility.CheckNotNull(nameof(sqlDialect), sqlDialect);
+
+    var storageTypeInfo = ColumnDefinition.StorageTypeInfo;
+
+    var parameterName = GetParameterName(sqlDialect);
+    var parameter = (SqlParameter)sqlDialect.CreateDataParameter(command, storageTypeInfo, parameterName, null);
+
+    // use the parameter we need anyway to find out the SqlDbType for the column
+    parameter.DbType = storageTypeInfo.StorageDbType;
+    var sqlMetaData = new SqlMetaData("Value", parameter.SqlDbType);
+
+    var typeName = $"TVP_{storageTypeInfo.StorageDbType}";
```

Comment:
Shouldn't we get the typename from a lookup/resolver?

Reply:
Eventually, yes. But that has been built in the context of RM-9216, which has not been merged yet - see https://github.com/re-motion/Framework/pull/1152
We will have to inject an `ISingleScalarStructuredTypeDefinitionProvider` into the `SqlDbCommandBuilderFactory` so that it can be injected into the `SqlTableValuedParameterColumnSpecification` objects. Then we can use `TableTypeDefinition.CreateTableValuedParameterValue()` here.
Perhaps we should do RM-9216 first, and make adjustments to this PR based on the new design?

Reply:
ah, okay. so, right now the code is locked on one specific pattern/usecase and the next PR fixes this. 

I can live with both orderings. officially cleaner would be to have the infrastructure first and then use the infrastructure here. but if thats' either more than 5min of work, it's probably not worth it? Was not sure if that was a bug or by design. perhaps add a comment for now if we keep the original order of PRs?